File: Day_17/singleTon/Singleton.java
package com.company.singleTon;

/**
 * 单例模式实现方式一： 立即加载/“饿汉模式”
 * 优点：实现简单，无多线程同步问题
 * 缺点：由于是静态变量，被创建后一直占用内存空间，某些情况下较为耗费内存。
 */
public class Singleton {
    // 自身实例化对象设置为一个属性，用static、final修饰
    private static final Singleton instance = new Singleton();

    // 构造方法私有化
    private Singleton(){}

    // 外部通过该静态方法调用实例
    public static Singleton getInstance(){
        return instance;
    }
}

File: 设计模式/单例模式/Singleton.go

type instance struct {
	name string
}

// 单例模式：“饿汉”模式

var lazyIns *instance

func GetLazyInstance() *instance {
	if lazyIns == nil {
		lazyIns = new(instance)
	}
	return lazyIns
}

/**
	"饿汉模式" 
 **/
var hunIns *instance

// 在包singleton包被导入后，init函数即刻自动运行
func init() {
	hunIns = new(instance)
	hunIns.name = "初始化单例模式：饿汉"
}
func GetHunInstance() *instance {
	return hunIns
}

// "双检查锁模式"
var mux sync.Mutex
func GetSafeInstance() *instance {
	if safeIns == nil {
		mux.Lock()
		defer mux.Unlock()
		if safeIns == nil {
			safeIns = new(instance)
		}
	}
	return safeIns
}

// sync包实现单例模式
var once sync.Once
var ins *instance

func GetInstance() *instance {
	once.Do(func() {
		ins = new(instance)
	})
	return ins
}

